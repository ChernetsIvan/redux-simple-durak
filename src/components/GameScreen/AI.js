import React from "react";
import PropTypes from "prop-types";
import { uniqueId } from "lodash-es";

import InvisibleCard from "./Card/InvisibleCard";
import HiddenCard from "./Card/HiddenCard";

const AI = ({ cards }) => {
  let output = null;
  if (cards.length > 0) {
    let inputCards = [];
    inputCards = cards.concat();
    let countOfCardsThatNotShown = 0; // сколько карт не видно
    let outCountOfCardsThatNotShown = null;
    // Выводим только первые 9 рубашкой кверху карт
    while (inputCards.length > 9) {
      inputCards.pop();
      countOfCardsThatNotShown += 1;
      outCountOfCardsThatNotShown = <b>+{countOfCardsThatNotShown}</b>;
    }

    const outputCards = inputCards.map(() => {
      const randKey = uniqueId();
      return <HiddenCard key={randKey} bootStrapColClass="col-1 ml-3" />;
    });

    output = (
      <div className="container mt-1">
        <div className="row justify-content-center">
          {outputCards}
          <div className="col-1 align-self-center">
            {outCountOfCardsThatNotShown}
          </div>
        </div>
      </div>
    );
  } else {
    output = <InvisibleCard />;
  }
  return (
    <div className="container">
      <div className="row">
        <div className="col-1" />
        <div className="col-10">{output}</div>
        <div className="col-1" />
      </div>
    </div>
  );
};

export default AI;

AI.propTypes = {
  cards: PropTypes.arrayOf(PropTypes.object).isRequired
};
