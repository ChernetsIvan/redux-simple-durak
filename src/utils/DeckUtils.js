import { pull, find } from 'lodash-es';

import {cardRanks, cardSuits} from './../constants/GameConstants';

import CardModel from './../models/CardModel';

class DeckUtils{
    //отсортировать входной массив карт по возрастанию isAsc=true
    //(убыванию-> isAsc=false)  по параметру 'power'
    static sortInputDeckByPower(inputDeck, isAsc){
        inputDeck.sort(function(a,b){
            if(isAsc){
                return a.power - b.power;
            }else{
                return b.power - a.power;
            }        
        });
    }

    //Возвращает наполненную всеми 36-ю картами колоду
    static fillCards(trumpSuit){
        let fullDeck = [];
        //Устанавливает power для каждой карты по принципу:
        //козырь->10*rank.cardValue, НЕкозырь->1*rank.cardValue
        let z = 0;
        let power, coef;               
        cardSuits.forEach(function(element,index, array){
            if(element.suit === trumpSuit.suit){
                coef = 10;
            }else{
                coef = 1;
            }
            cardRanks.forEach(function(element2, index2, array2){
                power = coef * element2.cardValue;
                fullDeck.push(
                    new CardModel(z, element2, element, power)
                );
                z++;
            });
        });
        return fullDeck;
    }

    //Возвращает перемешанную колоду
    static shuffle(inputDeck){
        var newFullDeck = [];
        while(inputDeck.length > 0){
            //Случайный индекс в inputDeck
            let randIndex = Math.random() * (inputDeck.length-1);
            randIndex = Math.round(randIndex);

            let card = inputDeck[randIndex];
            newFullDeck.push(card);

            inputDeck.splice(randIndex, 1);
        }
        return newFullDeck;
    }

    //Перемещает первую попавшуюся карту с козырной мастью
    //в конец колоды 
    static moveAnyCardWithTrumpSuitToTailOfFullDeck(trumpSuit, fullDeck){
        let card = find(fullDeck, (card) => {
            return card.suit.suit === trumpSuit.suit;
        });
        pull(fullDeck, card);
        fullDeck.push(card);
    }

    //взять из НАЧАЛА (!) колоды fullDeck и положить в inputDeck столько карт, 
    //чтобы в inputDeck в итоге было 6 карт
    static giveUpToSixCards(fullDeck, inputDeck){
        //если хотя бы одно из условий не выполняется, в цикл не заходим:
        while(inputDeck.length < 6 && fullDeck.length > 0){
            let card = fullDeck.shift();
            inputDeck.push(card);
        }
    }

    //Возвращает козырную масть (не карту!) для будущей игры
    static chooseSuitForTrumpCardInFuture(){
        let randIndex = Math.random() * (cardSuits.length-1);
        randIndex = Math.round(randIndex);
        return cardSuits[randIndex];
    }
}

export default DeckUtils;