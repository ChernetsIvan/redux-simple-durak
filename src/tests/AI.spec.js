import React from "react";
import renderer from "react-test-renderer";
import AI from "./../components/GameScreen/AI";
import DeckUtils from "./../utils/DeckUtils";

describe("AI snapshot testing", () => {

  test("AI recieve cards", () => {
    const cards = DeckUtils.fillCards({ suit: "Б" });
    const component = renderer.create(<AI cards={cards} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("AI don't recieve cards", () => {
    const cards = DeckUtils.fillCards({ suit: "Б" });
    const component = renderer.create(<AI cards={cards.slice(0,0)} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

});