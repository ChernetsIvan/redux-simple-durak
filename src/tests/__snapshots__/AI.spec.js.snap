// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AI snapshot testing AI don't recieve cards 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-1"
    />
    <div
      className="col-10"
    >
      <div
        className="container invisible"
      >
        <div
          className="row"
        >
          <div
            className="col-1 card mb-1 shirt"
          />
        </div>
      </div>
    </div>
    <div
      className="col-1"
    />
  </div>
</div>
`;

exports[`AI snapshot testing AI recieve cards 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-1"
    />
    <div
      className="col-10"
    >
      <div
        className="container mt-1"
      >
        <div
          className="row justify-content-center"
        >
          <div
            className="col-1 ml-3 card mb-1 shirt"
          />
          <div
            className="col-1 ml-3 card mb-1 shirt"
          />
          <div
            className="col-1 ml-3 card mb-1 shirt"
          />
          <div
            className="col-1 ml-3 card mb-1 shirt"
          />
          <div
            className="col-1 ml-3 card mb-1 shirt"
          />
          <div
            className="col-1 ml-3 card mb-1 shirt"
          />
          <div
            className="col-1 ml-3 card mb-1 shirt"
          />
          <div
            className="col-1 ml-3 card mb-1 shirt"
          />
          <div
            className="col-1 ml-3 card mb-1 shirt"
          />
          <div
            className="col-1 align-self-center"
          >
            <b>
              +
              27
            </b>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-1"
    />
  </div>
</div>
`;
